src/libraries/Common/src/System/Reflection/Metadata/TypeNameParserHelpers.cs

```diff
@@ -0,0 +1,377 @@
+            // Whitespace cannot precede the decorator, but it can follow the decorator.
+
+            ReadOnlySpan<char> originalInput = input; // so we can restore on 'false' return
+
+            if (TryStripFirstCharAndTrailingSpaces(ref input, '*'))
+            {
+                rankOrModifier = Pointer;
+                return true;
+            }
+
+            if (TryStripFirstCharAndTrailingSpaces(ref input, '&'))
+            {
+                rankOrModifier = ByRef;
+                return true;
+            }
+
+            if (TryStripFirstCharAndTrailingSpaces(ref input, '['))
+            {
+                // SZArray := []
+                // MDArray := [*] or [,] or [,,,, ...]
+
+                int rank = 1;
+                bool hasSeenAsterisk = false;
+
+            ReadNextArrayToken:
+
+                if (TryStripFirstCharAndTrailingSpaces(ref input, ']'))
+                {
+                    // End of array marker
+                    rankOrModifier = rank == 1 && !hasSeenAsterisk ? SZArray : rank;
+                    return true;
+                }
+
+                if (!hasSeenAsterisk)
+                {
+                    if (rank == 1 && TryStripFirstCharAndTrailingSpaces(ref input, '*'))
+                    {
+                        // [*]
+                        hasSeenAsterisk = true;
+                        goto ReadNextArrayToken;
+                    }
+                    else if (TryStripFirstCharAndTrailingSpaces(ref input, ','))
+                    {
+                        // [,,, ...]
+                        checked { rank++; }
+                        goto ReadNextArrayToken;
+                    }
+                }
+
+                // Don't know what this token is.
+                // Fall through to 'return false' statement.
+            }
+
+            input = originalInput; // ensure 'ref input' not mutated
+            rankOrModifier = 0;
+            return false;
+        }
+
+        internal static bool TryStripFirstCharAndTrailingSpaces(ref ReadOnlySpan<char> span, char value)
+        {
+            if (!span.IsEmpty && span[0] == value)
+            {
+                span = span.Slice(1).TrimStart();
+                return true;
+            }
+            return false;
+        }
+
+        internal static InvalidOperationException InvalidOperation_MaxNodesExceeded(int limit) =>
+#if SYSTEM_REFLECTION_METADATA || SYSTEM_PRIVATE_CORELIB
+            new InvalidOperationException(SR.Format(SR.InvalidOperation_MaxNodesExceeded, limit));
+#else // tools that reference this file as a link
+            new InvalidOperationException();
+#endif
+
+        internal static ArgumentException ArgumentException_InvalidTypeName(int errorIndex) =>
+#if SYSTEM_PRIVATE_CORELIB
+            new ArgumentException(SR.Arg_ArgumentException, $"typeName@{errorIndex}");
+#elif SYSTEM_REFLECTION_METADATA
+            new ArgumentException(SR.Argument_InvalidTypeName);
```

Comment:
Is it useful to include the position in the System.Reflection.Metadata error message as well? The position is very convenient when longer type names fail to parse.